llvm/lib/MC/MachObjectWriter.cpp

```diff
@@ -277,6 +277,8 @@ void MachObjectWriter::writeSection(const MCAssembler &Asm,
     W.write<uint32_t>(VMAddr);      // address
     W.write<uint32_t>(SectionSize); // size
   }
+  if (!isUInt<32>(FileOffset))
+    report_fatal_error("Cannot encode offset of section");
   W.write<uint32_t>(FileOffset);
 
   W.write<uint32_t>(Log2(Section.getAlign()));
```

Comment:
Mind adding another one for `RelocationStart`? Looks like it might have the same problem.

Reply:
We can call `Asm.getContext().reportError` instead to not trigger a crash.

Reply:
Perhaps let the caller (`MachObjectWriter::writeObject`) check this and early return.

Reply:
The point of the early return would be to avoid duplicate error messages?  That makes sense, I guess.

Reply:
Tried rearranging as suggested.